File: Leetcode/1038 Binary Search Tree to Greater Sum Tree.cpp
class Solution {
  public:
    TreeNode *bstToGst(TreeNode *root) {
        int greaterSum = 0;
        dfs(root, greaterSum);
        return root;
    }

    void dfs(TreeNode *root, int &greaterSum) {
        if (root == nullptr) {
            return;
        }
        dfs(root->right, greaterSum);
        greaterSum += root->val;
        root->val = greaterSum;
        dfs(root->left, greaterSum);
    }
};

File: Leetcode/1038 Binary Search Tree to Greater Sum Tree.py
class Solution:
    def bstToGst(self, root):
        greater_sum = 0

        def dfs(root):
            if root is None:
                return
            
            nonlocal greater_sum

            dfs(root.right)

            greater_sum += root.val
            root.val = greater_sum
            
            dfs(root.left)

        dfs(root)
        return root
